<center><h1> Assignment Solution 2</h1></center>
<center><h2>Manoj Dongare</h2></center>

#### Q1. What is a List?
<b>Ans - </b>List is a collection of items used to store sequence
of various data types(int, float, list, tuple, set, dict, dataframe)
Lists are mutable and can contains duplicates. These are ordered and are enclosed by square( [ ] ) brackets. <b>e.g.</b> list1 = [1, 2, 'c', ['c', 'd']]

#### Q2. What is a Tuple?
<b>Ans</b> - Tuples are same as lists and are immutable and are enclosed by round(( )) beackets. <b>e.g.</b> tuple1 = (2,3,4,5)

#### Q3. What is the difference between List and Tuple?
<b>Ans - </b> Lists are mutable and enclosed within square brackets, where as Tuples are immutable and are enclosed within round brackets.

#### Q4. Python Program to find the largest element in the list.

In [15]:
def findLargest(list1):
    largest = list1[0]
    for i in list1:
        if largest < i:
            largest = i
    return largest
        
print(findLargest([1,3,2,5,6,8]))

8


#### Q5. Python program to interchange first and last elements in a list.

In [11]:
def interchangeFIrstAndLast(list1):
    list1[0], list1[-1] = list1[-1], list1[0]
    return list1
print(interchangeFIrstAndLast([1,2,3,4,5,6,7,8,9]))
print(interchangeFIrstAndLast(['V','e','l','o','c','i','t','y']))

[9, 2, 3, 4, 5, 6, 7, 8, 1]
['y', 'e', 'l', 'o', 'c', 'i', 't', 'V']


#### Q6. Python program to swap two elements in a list.

In [14]:
def swapTwoElements(lst, index1, index2):
    lst[index1], lst[index2] = lst[index2], lst[index1]
    return lst
print(swapTwoElements([0,1,2,3,4,5,6,7,8,9], 7, 3))


[0, 1, 2, 7, 4, 5, 6, 3, 8, 9]


#### Q7. Python program to Reverse a List

In [16]:
def reverse(lst):
    return lst[::-1]
print(reverse('Velocity'))

yticoleV


#### Q8. Python program to count occurrences of an element in a list

In [4]:
def frequency(element, lst):
    return lst.count(element)
print(frequency(3,[1,3,2,3,5,6,3,6,8,9]))

3


#### Q9. Python program to find the sum of elements in a list

In [20]:
def SUM(lst):
    sum = 0;
    for element in lst:
        sum += element
    return sum
print(SUM([1,2,3,4,5,6,7,8,9]))

45


#### Q10. Python program to Multiply all numbers in the list

In [21]:
def MULTIPLY(lst):
    product = 1;
    for element in lst:
        product *= element
    return product
print(MULTIPLY([1,2,3,4,5,6,7,8,9]))

362880


#### Q11. What are the ways to find the length of a list

In [30]:
######## Method-1 #########
def len1(lst):
    count = 0
    for i in lst:
        count += 1
    return count

######## Method-2 #########
def len2(lst):
    count = 0
    for i in range(0, len(lst)):
        count += 1
    return count

######## Method-3 #########
def len3(lst):
    return len(lst)

print(len1([0,1,2,3,4,5,6,7,8,9]))
print(len2([0,1,2,3,4,5,6,7,8,9]))
print(len3([0,1,2,3,4,5,6,7,8,9]))


10
10
10


#### Q12. Python program to find the smallest and largest number in a list (Without min-max function)

In [33]:
def findSmallestAndLargest(list1):
    largest = list1[0]
    smallest = list1[0]
    for i in list1:
        if largest < i:
            largest = i
        if smallest > i:
            smallest = i
    return largest, smallest
        
print(findSmallestAndLargest([1,3,2,5,6,8]))

(8, 1)


#### Q13. Python Program to find the area of a circle

In [42]:
radius = 7
PI = 22/7
print('Area = ',PI*radius*radius)

Area =  22.07071428571428


#### Q14. Take inputs from the user to make a list. Again take one input from the user and search it in the list and delete that element, if found. Iterate over a list using for loop.

In [26]:
list1 = []
listSize = int(input('Enter the list size" '))
for i in range(0,listSize):
    list1.append(int(input('Enter values of List: ')))
    
num = int(input('Enter the number to search for" '))
for i in range(len(list1)-1):
    if list1[i] == num:
        list1.remove( num)
print(f'List after deleting {num}:', list1)

Enter the list size" 4
Enter values of List: 1
Enter values of List: 2
Enter values of List: 3
Enter values of List: 4
Enter the number to search for" 3
List after deleting 3: [1, 2, 4]


#### Q15. You are given a list of integer elements. Make a new list that will store a
square of elements of the previous list. (With and without list comprehension)  
i. Input_list = [2,5,6,12]  
ii. Output_list = [4,25,36,144]  

In [29]:
def withComprehension(lst):
    return [i*i for i in lst]
Input_list = [2,5,6,12]

def withOutComprehension(lst):
    result = []
    for i in lst:
        result.append(i*i)
    return result
Input_list = [2,5,6,12]

print('With comprehension   ', withComprehension(Input_list))
print('Without comprehension', withOutComprehension(Input_list))


With comprehension    [4, 25, 36, 144]
Without comprehension [4, 25, 36, 144]


#### Q16. WAP to create two lists, one containing all even numbers and the other containing all odd numbers between 0 to 151

In [35]:
even = []
odd  = []
for i in range(152):
    if i%2 == 0:
        even.append(i)
    else:
        odd.append(i)
print('Even numbers: ', even)
print('Odd numbers : ', odd)

Even numbers:  [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150]
Odd numbers :  [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151]


#### Q17. Python program to Count Even and Odd numbers in a List

In [38]:
def countEvenOdd(list1):
    even = 0
    odd  = 0
    for i in list1:
        if i%2 == 0:
            even += 1
        else:
            odd += 1
    print('Even numbers: ', even)
    print('Odd numbers : ', odd)
countEvenOdd([2,4,7,8,3,5,6,35,73,35])

Even numbers:  4
Odd numbers :  6


#### Q18. WAP to make new lists, containing only numbers which are divisible by 4, 6, 8, 10, 3, 5, 7, and 9 in separate lists for range(0,151)

In [6]:
res3 = []
res4 = []
res5 = []
res6 = []
res7 = []
res8 = []
res9 = []
res10 = []
for i in range(152):
    if i%3==0:
        res3.append(i)
    if i%4==0:
        res4.append(i)
    if i%5==0:
        res5.append(i)
    if i%6==0:
        res6.append(i)
    if i%7==0:
        res7.append(i)
    if i%8==0:
        res8.append(i)
    if i%9==0:
        res9.append(i)
    if i%10==0:
        res10.append(i)

print('res3', res3)
print('res4', res4)
print('res5', res5)
print('res6', res6)
print('res7', res7)
print('res8', res8)
print('res9', res9)
print('res10', res10)

res3 [0, 3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99, 102, 105, 108, 111, 114, 117, 120, 123, 126, 129, 132, 135, 138, 141, 144, 147, 150]
res4 [0, 4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 72, 76, 80, 84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124, 128, 132, 136, 140, 144, 148]
res5 [0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100, 105, 110, 115, 120, 125, 130, 135, 140, 145, 150]
res6 [0, 6, 12, 18, 24, 30, 36, 42, 48, 54, 60, 66, 72, 78, 84, 90, 96, 102, 108, 114, 120, 126, 132, 138, 144, 150]
res7 [0, 7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98, 105, 112, 119, 126, 133, 140, 147]
res8 [0, 8, 16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96, 104, 112, 120, 128, 136, 144]
res9 [0, 9, 18, 27, 36, 45, 54, 63, 72, 81, 90, 99, 108, 117, 126, 135, 144]
res10 [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150]


#### Q19. From a list containing ints, strings, and floats, make three lists to store them separately.

In [3]:
def dTypeSeparator(lst):
    intList = []
    floatList = []
    strList = []
    for i in lst:
        if type(i) == int:
            intList.append(i)
        if type(i) == float:
            floatList.append(i)
        if type(i) == str:
            strList.append(i)
    print('Integers: ', intList)
    print('Floats', floatList)
    print('Strings', strList)
dTypeSeparator([1, 'Python', 3.14])

Integers:  [1]
Floats [3.14]
Strings ['Python']


#### Q20. What’s The Difference Between The Python append() and extend() Methods?

<font color = blue><b>Append</b></font> function Adds the argument as single list. Increases size of list by one.  
<b>e.g.</b>  
list1 = [1, 2, 3]  
list2 = [4, 5, 6]  
list1.append(list2)  
Here list1 will contain [1, 2, 3, [4, 5, 6] ]  
<hr width = 7/>
Where as <font color = blue><b>Extend</b></font> will iterate over its argument and adding each element, extending the list.<br>
<b>e.g.</b><br>
list1.append(list2)<br>  
Here list2 will contain elements as [1, 2, 3, 4, 5, 6]

#### Q21. Write a Python program to append a list to the second list

In [1]:
l1 = [1,2,3]
l2 = [4,5,6]
l1.append(l2)
print(l1)

[1, 2, 3, [4, 5, 6]]


#### Q22. Write a Python program to find the third-largest number in a list

In [5]:
a = [3,55,34,67,84,23,7,90,45,64]
a.sort()
print(a[-3])

67


#### Q23. Write a Python program to get the frequency of the elements in a list.

In [8]:
def frequency(element, list1):
    return list1.count(element)
print(frequency(3, [1,4,3,2,3,5,6,7,3,5,3]))

4


#### Q24. Write a Python program to check whether a list contains a sublist

In [15]:
def SubList(sublist, mainlist):
    for element in sublist:
        if element not in mainlist:
            return False
    return True
print(SubList([1,3,5],[1,2,3,4,5,6,7,8,9]))
print(SubList([1,3,5,76],[1,2,3,4,5,6,7,8,9]))


True
False


#### Q25. Write a Python program to generate all sublists of a list

In [ ]:
lst = ['a','b','c']
for i in range(len(lst)):
    for j in range(len(lst)):
        

#### Q26. Write a Python program to find common items from two lists

In [3]:
def common(l1,l2):
    result = []
    for i in l1:
        for j in l2:
            if i == j:
                result.append(i)
    return result
print(common([1,2,3], [2,3,4]))

[2, 3]


#### Q27. How to flatten a list in python?

In [4]:
def flatten(unflattenedList):
    flattenedList = []
    for list1 in unflattenedList:
        for element in list1:
            flattenedList.append(element)
    return flattenedList
print(flatten([[1, 2, 3], [4, 5, 6], [7], [8, 9]]))    

[1, 2, 3, 4, 5, 6, 7, 8, 9]


#### Q28. How to sort a list in ascending and descending order without using the sort function?

In [14]:
def sort(lst):
    length = len(lst)
    for i in range(length):
        for j in range(length):
            if lst[i] <= lst[j]:
                lst[i],lst[j] = lst[j],lst[i]
    return lst
new = sort([34,67,23,96,12,56])
print('Ascending order is: ',new)
print('Descending order is: ',new[::-1])

Ascending order is:  [12, 23, 34, 56, 67, 96]
Descending order is:  [96, 67, 56, 34, 23, 12]


#### Q29. How to sort a tuple?

In [15]:
tuple1 = (3,2,5,3,7,2,9,4,3,2)
list1 = sorted(tuple1)
print(list1)

[2, 2, 2, 3, 3, 3, 4, 5, 7, 9]

#### Q30. Write a Python program to convert a list of multiple integers into a single integer  
a. [11, 33, 50] >>> 113350

In [1]:
def multiIntToSinglInt(value):
    result = ''
    for num in value:
        result += str(num)
    return result
     
print(multiIntToSinglInt([11, 33, 50]))

113350


In [ ]:
31. Difference between del and clear?

| del  || clear    |
|:-    ||:-        |
| -Is a keyword  || -Is a function       |
| -Deletes list object  || -Empties list    |
|<b>-Syntax- </b> del list <b>or</b> del list[ i ]||<b>-Syntax- </b>list.clear|

#### Q32. Difference between remove and pop?

|remove()  ||pop()    |
|:-    ||:-        |
|-Can remove any value||-Removes last value by default|
|-Takes value as parameter||-Takes index as parameter|
|-No return value||-Returns deleted value|

#### Q33. Difference between indexing and Slicing

| indexing || slicing|
|:-||:-|
|-Used to obtain individual items||-Used to obtain sequence of items|
|-Array out of index error||-No error for out of index access|
|-Returns single element||-May return new sequence|
||||

#### Q34. Difference between sort and sorted?

|sort()||sorted()|
|:-||:-|
|-Sorts original list||-Returns new sorted list|
|-Is a method||-Is a built in function|
|-Can sort only list||-Can sort any sequence|
|-Does not return anything||-Returns new sorted sequence|

#### Q35. Difference between reverse and reversed?

|reverse()||reversed()|
|:-||:-|
|-Sorts original list||-Returns new sorted list|
|-Can apply on list only||-Can apply on any sequence|
|-Does not return anything||-Returns new reversed sequence|

#### Q36. Difference between copy and deep copy?

|copy||deep copy|
|:-||:-|
|-Creates new object which stores the reference of original element||-Creates new object and recursively adds the copies of nested objects present in the original object|
|-If we change any element, it will reflect to original also||-If we change any element in copied object, it will not reflect original object|

#### Q37. How to check whether the list is empty or not?

In [1]:
list1 = []
####### FIrst method #####
if len(list1) == 0:
    print('List is empty')
##### Second method #####
if not list1:
    print('List is empty')

List is empty
List is empty


#### Q38. How to concatenate two lists?

In [3]:
l1 = [1,2,3,4]
l2 = ['a','b','c']
#First way
l3 = l1+l2
print(l3)
#Second way
l1.extend(l2)
print(l1)

[1, 2, 3, 4, 'a', 'b', 'c']
[1, 2, 3, 4, 'a', 'b', 'c']


#### Q39. How to find the occurrences of an element in the python list?

In [ ]:
def frequency(element, lst):
    return lst.count(element)
print(frequency(3,[1,3,2,3,5,6,3,6,8,9]))

#### Q40. How to flatten a list in python?

In [ ]:
def flatten(unflattenedList):
    flattenedList = []
    for list1 in unflattenedList:
        for element in list1:
            flattenedList.append(element)
    return flattenedList
print(flatten([[1, 2, 3], [4, 5, 6], [7], [8, 9]]))